push-messaging-and-notifications/main.js

```diff
@@ -0,0 +1,175 @@
+
+          pushButton.disabled = false;
+          pushButton.textContent = 'Enable Push Messages';
+          isPushEnabled = false;
+
+          if (!successful) {
+            console.error('We were unable to unregister from push');
+            return;
+          }
+        }).catch(function(e) {
+          // We failed to unsubscribe, this can lead to
+          // an unusual state, so may be best to remove 
+          // the users data from your data store and 
+          // inform the user that you disabled push
+
+          console.log('Unsubscription error: ', e);
+          pushButton.disabled = false;
+        });
+      }).catch(function(e) {
+        console.error('Error thrown while unsubscribing from push messaging.', e);
+      });
+  });
+}
+
+
+function subscribeToPushManager() {
+  // Disable the button so it can't be changed while
+  // we process the permission request
+  var pushButton = document.querySelector('.js-push-button');
+  pushButton.disabled = true;
+
+  navigator.serviceWorker.ready.then(function(serviceWorkerRegistration) {
+    serviceWorkerRegistration.pushManager.subscribe()
+      .then(function(subscription) {
+        // The subscription was successful
+        isPushEnabled = true;
+
+        // TODO: Send the subscriptionId and endpoint to your server
+        // and save it to send a push message at a later date
+        var subscriptionId = subscription.subscriptionId;
+        var endpoint = subscription.endpoint;
+        sendToServer(subscriptionId, endpoint);
+        
+        pushButton.textContent = 'Disable Push Messages';
+        pushButton.disabled = false;
+      })
+      .catch(function(e) {
+        if (Notification.permission === 'denied') {
+          // The user denied the notification permission which
+          // means we failed to subscribe and the user will need
+          // to manually change the notification permission to
+          // subscribe to push messages
+          console.warn('Permission for Notifications was denied');
+          pushButton.disabled = true;
+        } else {
+          // A problem occurred with the subscription, this can
+          // often be down to an issue or lack of the gcm_sender_id
+          // and / or gcm_user_visible_only
+          console.error('Unable to subscribe to push.', e);
+          pushButton.disabled = false;
+        }
+      });
+  });
+}
+
+// Once the service worker is registered set the initial state
+function initialiseState() {
+  // Are Notifications supported?
+  if (!('Notification' in window)) {
+    console.warn('Notifications aren\'t supported.');
+    return;
+  }
+
+  // Check the current Notification permission.
+  // If its denied, it's a permanent block until the
+  // user changes the permission
+  if (Notification.permission === 'denied') {
+    console.warn('The user has blocked notifications.');
+    return;
+  }
+
+  // Check if push messaging is supported
+  if (!('PushManager' in window)) {
```

Comment:
I wasn't aware that `window.PushManager` was a thing. Are you checking for that as a proxy, since it's more convenient than getting a `ServiceWorkerRegistration` and checking for `pushManager`? Or is there a real reason that this code won't work if `window.PushManager` is not defined?